R/list-simplify.R

```diff
@@ -0,0 +1,89 @@
+#' Simplify a list to an atomic or S3 vector
+#'
+#' @details
+#' Simplification maintains a one-to-one correspondence between the input
+#' and output, implying that each element of `x` must contain a vector of
+#' length 1. If you don't want to maintain this correspondence, then you
+#' probably want either [list_c()] or [list_flatten()].
+#'
+#' @param x A list.
+#' @param strict What should happen if simplification fails? If `TRUE`,
+#'   will error. If `FALSE`, will return `x` unchanged.
+#' @param ptype An optional prototype to ensure that the output type is always
+#'   the same.
+#' @returns A vector the same length as `x`.
+#' @export
+#' @examples
+#' list_simplify(list(1, 2, 3))
+#'
+#' try(list_simplify(list(1, 2, "x")))
+#' try(list_simplify(list(1, 2, 1:3)))
+list_simplify <- function(x, strict = TRUE, ptype = NULL) {
+  simplify_impl(x, strict = strict, ptype = ptype)
+}
+
+# Wrapper used by purrr functions that do automatic simplification
+list_simplify_internal <- function(
+    x,
+    simplify = NA,
+    ptype = NULL,
+    error_arg = "x",
+    error_call = caller_env()
+  ) {
+  if (length(simplify) > 1 || !is.logical(simplify)) {
+    cli::cli_abort("{.arg simplify} must be `TRUE`, `FALSE`, or `NA`.")
+  }
+  if (!is.null(ptype) && isFALSE(simplify)) {
+    cli::cli_abort("Must not specify {.arg ptype} when `simplify = FALSE`.")
+  }
+
+  if (isFALSE(simplify)) {
+    return(x)
+  }
+
+  simplify_impl(
+    x,
+    strict = !is.na(simplify),
+    ptype = ptype,
+    error_arg = error_arg,
+    error_call = error_call
+  )
+}
+
+simplify_impl <- function(
+    x,
+    strict = TRUE,
+    ptype = NULL,
+    error_arg = "`x`",
+    error_call = caller_env()
+  ) {
+  vec_check_list(x, arg = error_arg, call = error_call)
+
+  can_simplify <- every(x, vec_is, size = 1)
+
+  if (can_simplify) {
+    tryCatch(
+      vec_unchop(x, ptype = ptype),
+      vctrs_error_incompatible_type = function(err) {
+        if (strict || !is.null(ptype)) {
+          cli::cli_abort(
```

Comment:
I'd consider `vec_unchop()` an internal detail so it's probably better to zap its error call:

```r
err$call <- NULL
```

If it had an `error_call` argument, we could zap it there instead, but it hasn't yet.

Reply:
Actually I think that's a case where we'd just take ownership of `vec_unchop()` errors instead of creating a chained error? So we'd do something like:

```r
vec_unchop(x, ptype = ptype, error_call = current_env())
```

Reply:
Yeah, I couldn't quite figure out how to get all the errors to be consistent here. I'll take another pass.